src/network/networkd-dhcp4.c

```diff
@@ -442,6 +443,12 @@ static int dhcp_lease_acquired(sd_dhcp_client *client, Link *link) {
                         r = link_set_hostname(link, hostname);
                         if (r < 0)
                                 log_link_error_errno(link, r, "Failed to set transient hostname to '%s': %m", hostname);
+                        else if (r == 0) {
+                                if (free_and_strdup(&link->manager->pending_hostname, hostname) < 0)
+                                    return log_oom();
+                                link->manager->pending_hostname_link = link;
```

Comment:
if the link disappears after this then the pending_hostname_link will point to nowhere, and we might deref it and access invalid memory...